c2rust-analyze/src/rewrite/span_index.rs

```diff
@@ -18,14 +19,26 @@ impl<T> SpanIndex<T> {
 
     /// Iterate over items whose spans are entirely contained within `span`.  The order of the
     /// returned items is unspecified.
-    pub fn lookup(&self, span: Span) -> RangeIter<T> {
+    pub fn _lookup(&self, span: Span) -> RangeIter<T> {
         let data = span.data();
         let start = self.v.partition_point(|(span, _)| span.lo() < data.lo);
         RangeIter {
             inner: self.v[start..].iter(),
             hi: data.hi,
         }
     }
+
+    /// Iterate over items whose spans are exactly equal to `span`.  The order of the returned
+    /// items is unspecified.
+    pub fn lookup_exact(&self, span: Span) -> ExactIter<T> {
+        let data = span.data();
+        let start = self.v.partition_point(|(span, _)| span.lo() < data.lo);
+        ExactIter {
+            inner: self.v[start..].iter(),
+            lo: data.lo,
+            span: span,
+        }
+    }
```

Comment:
If we return an `impl Iterator` and use iterator methods, we don't need `ExactIter`, and I think the iterator chaining way is a lot easier to understand.

```suggestion
    pub fn lookup_exact(&self, span: Span) -> impl Iterator<Item = &T> {
        let data = span.data();
        let start = self.v.partition_point(|(span, _)| span.lo() < data.lo);
        self.v[start..]
            .iter()
            // The list is sorted by `lo`, so once we move past `self.lo`, we won't see it again.
            .take_while(move |(span, _)| span.data().lo == data.lo)
            // This is not the requested span.  Note we compare the entire span, not just `lo`
            // and `hi`, because spans with the same `lo` and `hi` can come from two different
            // macro expansions (thus having different `SyntaxContext`s).
            .filter(move |(span_, _)| *span_ == span)
            .map(|(_, t)| t)
    }
```

Reply:
in either of these implementations why is it necessary to partition and cut off iteration? it seems like filtering by equality with span would optimize for clarity at the (maybe minor?) expense of iterating past the span's lo byte

Reply:
I think the expense is actually significant here.  Using `filter` instead of breaking early means visiting roughly every span in the function on every lookup, making the overall behavior O(N^2).  Since we know the list is sorted, we can do better by looking only at the range that could plausibly contain matching entries.

Reply:
That sounds fine but, just for my own understanding, isn't the behavior O(N^2) anyway because of the edge case where all spans match the criteria?

Reply:
Yes, in the worst case it's still O(N^2), but in most cases we expect to see, it should be closer to O(N).

Reply:
I switched to iterator chains here and in `fn lookup()`.